core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.rpc;
+
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Scope;
+
+public class AccumuloProtocolFactory extends TCompactProtocol.Factory {
+
+  private static final long serialVersionUID = 1L;
+
+  private final boolean isClient;
+
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    private static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    private static final byte PROTOCOL_VERSION = 1;
+
+    private final boolean isClient;
+
+    private Span span = null;
+    private Scope scope = null;
+
+    public AccumuloProtocol(TTransport transport, boolean isClient) {
+      super(transport);
+      this.isClient = isClient;
+    }
+
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (this.isClient) {
+        span = TraceUtil.startClientRpcSpan(this.getClass(), message.name);
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
+        } catch (TException e) {
+          if (scope != null) {
+            scope.close();
+          }
+          if (span != null) {
+            span.end();
+          }
+          throw e;
+        }
+      }
+
+      super.writeMessageBegin(message);
```

Comment:
Seems like this should be inside the try block above.  Thinking if this message throws exception that writeMessageEnd will never be called.  If it should be in the try block then maybe the existing code in 2.1 has a bug.

Reply:
I'm not sure I understand. Putting `super.writeMessageBegin(message);` in the try-catch would close the span and scope if an exception is thrown from it but I'm not sure that it would make `writeMessageEnd()` in either case, right? Maybe im missing something here.

Reply:
If `super.writeMessageBegin(message)` throws an exception, then it we can probably assume that `writeMessageEnd()` will never be called. So if this happens will nothing close the span and scope?

Reply:
Oh yea, that sounds correct. I misunderstood what you were saying. I'll make sure span and scope get closed in that case.

Reply:
Should be fixed as of 1ea49cc